{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### This code has been developed and run using the biodesign_3.7 kernel on the jprime.lbl.gov server. It uses the cplex library for running the MOMA optimization"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import collections as col\n",
    "import os\n",
    "import random\n",
    "import re\n",
    "import sys\n",
    "import warnings\n",
    "from shutil import copyfile\n",
    "from enum import Enum\n",
    "from typing import NewType, Dict, List, Any, OrderedDict, Counter\n",
    "\n",
    "import cobra\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from cobra.exceptions import OptimizationError, Infeasible"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Global parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Type annotations\n",
    "Filename = NewType('Filename', str)\n",
    "\n",
    "# Enumerations\n",
    "class Omics(Enum):\n",
    "    \"\"\"Enumeration with supported omics data types.\"\"\"\n",
    "    PROTEOMICS = 0\n",
    "    TRANSCRIPTOMICS = 1\n",
    "    METABOLOMICS = 2\n",
    "\n",
    "    def __str__(self):\n",
    "        return f'{str(self.name).lower()}'\n",
    "\n",
    "\n",
    "# Constants\n",
    "UNIPROT_URL = '''https://www.uniprot.org/uploadlists/'''\n",
    "CTS_URL = '''https://cts.fiehnlab.ucdavis.edu/rest/convert/'''\n",
    "# HOST NAME\n",
    "HOST_NAME: str = 'ropacus' \n",
    "# TODO: Move some constants to variables by program arguments\n",
    "DATA_FILE_PATH: Filename = Filename('data')\n",
    "# Output file path\n",
    "OUTPUT_FILE_PATH: Filename = Filename('data/output')\n",
    "# INCHIKEY_TO_CID_MAP_FILE_PATH: mapping file path to map inchikey to cids\n",
    "INCHIKEY_TO_CID_MAP_FILE_PATH: Filename = Filename('mapping') \n",
    "# MODEL_FILENAME: Filename = Filename('iECIAI39_1322.xml')  # E. coli\n",
    "MODEL_FILENAME: Filename = Filename('reannotated_base_v3.sbml')  # R. opacus\n",
    "# TODO: rename traininig file name to design file names and paths\n",
    "# Training file name\n",
    "TRAINING_FILE_NAME: Filename = Filename('')\n",
    "TRAINING_FILE_PATH: Filename = Filename('')\n",
    "# Start time and stop time\n",
    "TIMESTART: float = 0.0\n",
    "TIMESTOP: float = 8.0\n",
    "NUMPOINTS: int = 9\n",
    "# number of reactions and instances\n",
    "NUM_REACTIONS: int = None\n",
    "NUM_INSTANCES: int = None\n",
    "\n",
    "# NOTE: user input to the program\n",
    "BIOMASS_REACTION_ID: str = 'BIOMASS_Ec_iJO1366_core_53p95M'  # E. coli\n",
    "REACTION_ID: str = 'biomass_target'  # R. opacus\n",
    "# REACTION_ID: str = 'SRC_C00185_e'  # R. opacus\n",
    "GENE_IDS_DBS: List[str] = ['kegg.genes']  # R. opacus\n",
    "# GENE_IDS_DBS: List[str] = ['uniprot', 'goa', 'ncbigi']  # E. coli\n",
    "UNITS: Dict[Omics, str] = {\n",
    "    Omics.PROTEOMICS: 'proteins/cell',\n",
    "    Omics.TRANSCRIPTOMICS: \"FPKM\",\n",
    "    Omics.METABOLOMICS: \"mg/L\"\n",
    "}\n",
    "# Fix the flux value to a particular value\n",
    "LOWER_BOUND: int = -15\n",
    "UPPER_BOUND: int = -15"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## FUNCTION A\n",
    "### [Fluxes,extMetabolitesCon, OD]   = getFluxTimeSeries(GSM, InitialExtMetabolitesCon)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_flux_time_series(model, ext_metabolites, parameters):\n",
    "    '''\n",
    "    Generate fluxes and OD\n",
    "    '''\n",
    "    \n",
    "   \n",
    "    # time steps to calculate the biomass production for\n",
    "    t0 = parameters['timestart']\n",
    "    tf = parameters['timestop']\n",
    "    points = parameters['numpoints']\n",
    "    tspan, delt = np.linspace(t0, tf, points, dtype='float64', retstep=True)\n",
    "\n",
    "    # panda series containing OD values for the timepionts\n",
    "    cell = pd.Series(index=tspan)\n",
    "    # NOTE: Do we to define this by the user?\n",
    "    cell0 = user_params['initial_OD'] # in gDW/L\n",
    "    cell[t0] = cell0\n",
    "\n",
    "    # getting external metabolites and their initial concentrations\n",
    "    met_names = []\n",
    "    initial_concentrations = []\n",
    "    for met, init_conc in ext_metabolites.items():\n",
    "        met_names.append(met)\n",
    "        initial_concentrations.append(init_conc)\n",
    "\n",
    "    # Dataframe containing external metabolites\n",
    "    subs = pd.DataFrame(index=tspan, columns=met_names)\n",
    "    # initial concentrations of external metabolites\n",
    "    subs.loc[t0] = initial_concentrations\n",
    "    \n",
    "    # fluxomics data for all timepoints\n",
    "    fluxomics_timeseries = {}\n",
    "    \n",
    "    # series to contain the biomass optimization solution across timepoints\n",
    "    sol_time_wild = pd.Series(index=tspan)\n",
    "\n",
    "    # exchange reactions and external metabolites dictionary\n",
    "    subs_ext = {r.id: r.reactants[0].id for r in model.exchanges if r.reactants[0].id in met_names}\n",
    "\n",
    "    # NOTE: put the body of the for loop inside a function\n",
    "    for t in tspan:\n",
    "        # Not changing the model but adding constraints for each time point\n",
    "        with model:\n",
    "            # why do we set volume to one? Is it arbitrary?\n",
    "            volume = 1.0\n",
    "            for k, v in subs_ext.items():\n",
    "                # Set global reactions bounds (in addition to local)\n",
    "                # set the lower bound to the maximum of the lower bound of the metabolite in the model \n",
    "                # and the existing amount of the metabolite so that it doe not go below the lower bound\n",
    "                model.reactions.get_by_id(k).lower_bound = max(model.reactions.get_by_id(k).lower_bound, -subs.loc[t,v]*volume/cell[t]/delt)\n",
    "            \n",
    "            # get optimized solution later used to calculate metabolite concentrations and fluxes in model\n",
    "            sol_t = model.optimize()\n",
    "            \n",
    "            # Step 2: storing the solution for each timepoint which are going to be reference solutions for moma\n",
    "            sol_time_wild[t] = sol_t\n",
    "            \n",
    "            # Step 3: Calculate mu where mu = solution(biomass)\n",
    "            mu = sol_t[BIOMASS_REACTION_ID]\n",
    "            print(t, sol_t.status, mu)\n",
    "            \n",
    "            # Step 4: Calculate ext. metabolite concentrations for next time point t+delta:\n",
    "#             iso = 'EX_isoprenol_e'\n",
    "            if sol_t.status == 'optimal' and mu > 1e-6:\n",
    "                # Calculating next time point's OD\n",
    "                cell[t+delt] = cell[t]*np.exp(mu*delt)\n",
    "                # Calculating external substrate concentrations for next time point\n",
    "                for k, v in subs_ext.items():\n",
    "                    subs.loc[t+delt,v] = max(subs.loc[t,v]-sol_t[k]/mu*cell[t]*(1-np.exp(mu*delt)),0.0)     \n",
    "#                     if k == iso:\n",
    "#                         if sol_t[k] > 0:\n",
    "#                             # Calculating isoprenol concentration for next time point\n",
    "#                             subs.loc[t+delt,v] = subs.loc[t,v]-sol_t[k]/mu*cell[t]*(1-np.exp(mu*delt))\n",
    "#                         else:\n",
    "#                             subs.loc[0:t, v] = 0\n",
    "#                             subs.loc[t+delt,v] = subs.loc[t,v]-sol_t[k]/mu*cell[t]*(1-np.exp(mu*delt))\n",
    "#                     else:\n",
    "#                         subs.loc[t+delt,v] = max(subs.loc[t,v]-sol_t[k]/mu*cell[t]*(1-np.exp(mu*delt)),0.0)     \n",
    "                \n",
    "            else:\n",
    "                cell[t+delt] = cell[t]\n",
    "                for k, v in subs_ext.items():\n",
    "                    subs.loc[t+delt,v] = subs.loc[t,v]\n",
    "\n",
    "            # get fluxomics data from the GSM and\n",
    "            # appending the fluxomics data for each timepoint into a Series\n",
    "            fluxomics_timeseries[t] = sol_t.fluxes.to_dict()\n",
    "            \n",
    "    return fluxomics_timeseries, cell, subs, subs_ext, sol_time_wild"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Function C\n",
    "### (Trans,Prot,Met) = getMultiomics(Fluxes, GSM)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_pubchem_id_file():\n",
    "    inchikey_to_cid = {}\n",
    "    filename = f'{INCHIKEY_TO_CID_MAP_FILE_PATH}/inchikey_to_cid.txt'\n",
    "    with open(filename, 'r') as fh:\n",
    "            try:\n",
    "                line = fh.readline()\n",
    "                while line:\n",
    "                    # checking to ignore inchikey records with no cid mappings\n",
    "                    if (len(line.split()) > 1):\n",
    "                        inchikey_to_cid[line.split()[0]] = 'CID:'+line.split()[1]\n",
    "                    else:\n",
    "                        inchikey_to_cid[line.strip()] = None\n",
    "\n",
    "                    line = fh.readline()\n",
    "            # NOTE: propagated exception, raise\n",
    "            except Exception as ex:\n",
    "                print(\"Error in reading file!\")\n",
    "                print(ex)\n",
    "\n",
    "    return inchikey_to_cid\n",
    "\n",
    "\n",
    "def get_proteomics_transcriptomics_data(model, solution):\n",
    "    \"\"\"\n",
    "\n",
    "    :param model:\n",
    "    :param solution:\n",
    "    :param condition:\n",
    "    :return:\n",
    "    \"\"\"\n",
    "\n",
    "    # pre-determined linear constant (NOTE: Allow user to set this via parameter)\n",
    "    # DISCUSS!!\n",
    "    k = 0.8\n",
    "    q = 0.06\n",
    "\n",
    "    proteomics = {}\n",
    "    transcriptomics = {}\n",
    "\n",
    "    rxnIDs = solution.fluxes.keys()\n",
    "    for rxnId in rxnIDs:\n",
    "        reaction = model.reactions.get_by_id(rxnId)\n",
    "        for gene in list(reaction.genes):\n",
    "\n",
    "            # this will ignore all the reactions that does not have the gene.annotation property\n",
    "            # DISCUSS!!\n",
    "            if gene.annotation:\n",
    "                if 'uniprot' not in gene.annotation:\n",
    "                    if 'goa' in gene.annotation:\n",
    "                        protein_id = gene.annotation['goa']\n",
    "                    else:\n",
    "                        break\n",
    "                else:\n",
    "                    protein_id = gene.annotation['uniprot'][0]\n",
    "                \n",
    "                # add random noise wjhich is 5 percent of the signal\n",
    "                noiseSigma = 0.05 * solution.fluxes[rxnId]/k;\n",
    "                noise = noiseSigma*np.random.randn();\n",
    "                proteomics[protein_id] = (solution.fluxes[rxnId]/k) + noise\n",
    "\n",
    "            # create transcriptomics dict\n",
    "            noiseSigma = 0.05 * proteomics[protein_id]/q;\n",
    "            noise = noiseSigma*np.random.randn();\n",
    "            transcriptomics[gene.id] = (proteomics[protein_id]/q) + noise\n",
    "\n",
    "    return proteomics, transcriptomics\n",
    "\n",
    "# NOTE: Work on this\n",
    "def get_metabolomics_data(model):\n",
    "    \"\"\"\n",
    "\n",
    "    :param model:\n",
    "    :param condition:\n",
    "    :return:\n",
    "    \"\"\"\n",
    "    metabolomics = {}\n",
    "    # get metabolites\n",
    "\n",
    "    # read the inchikey to pubchem ids mapping file\n",
    "    inchikey_to_cid = {}\n",
    "    inchikey_to_cid = read_pubchem_id_file()\n",
    "    \n",
    "    for met in model.metabolites:\n",
    "        # get associated reactions involving metabolite\n",
    "        num_reactions_with_flux = 0\n",
    "        for reaction in list(met.reactions):\n",
    "            # get dictionary of associated metabolites and their concentrations\n",
    "            # average fluxes\n",
    "            \n",
    "            # if there is an inchikey ID for the metabolite\n",
    "            if 'inchi_key' in met.annotation:\n",
    "                # if it is a list get the first element\n",
    "                if type(met.annotation['inchi_key']) is list:\n",
    "                    inchi_key = met.annotation['inchi_key'][0]\n",
    "                else:\n",
    "                    inchi_key = met.annotation['inchi_key']\n",
    "                \n",
    "                # if the inchikey for the metabolite is in the mapping dict\n",
    "                if inchi_key in inchikey_to_cid.keys():\n",
    "                    # if the CID is not in the metabolomics dict keys AND the mappned value is not None and the reactions flux is not 0\n",
    "                    if (inchikey_to_cid[inchi_key] not in metabolomics.keys()) and (inchikey_to_cid[inchi_key] is not None) and reaction.flux != 0.0:\n",
    "                            metabolomics[inchikey_to_cid[inchi_key]] = abs(reaction.flux)\n",
    "                            num_reactions_with_flux  += 1\n",
    "                    elif (inchikey_to_cid[inchi_key] is not None) and reaction.flux != 0.0:\n",
    "                            metabolomics[inchikey_to_cid[inchi_key]] += abs(reaction.flux)\n",
    "                            num_reactions_with_flux  += 1\n",
    "\n",
    "        # check if inchi_key attribite present else ignore metabolite\n",
    "        if 'inchi_key' in met.annotation.keys() and inchi_key in inchikey_to_cid.keys() and inchikey_to_cid[inchi_key] is not None:\n",
    "            metabolomics[inchikey_to_cid[inchi_key]]/=num_reactions_with_flux\n",
    "\n",
    "    return metabolomics\n",
    "\n",
    "def get_multiomics(model, solution):\n",
    "    \"\"\"\n",
    "\n",
    "    :param model: cobra model object\n",
    "    :param solution: solution for the model optimization using cobra\n",
    "    :param data_type: defines the type of -omics data to generate (all by default)\n",
    "    :return:\n",
    "    \"\"\"\n",
    "\n",
    "    proteomics = {}\n",
    "    transcriptomics = {}\n",
    "    fluxomics = {}\n",
    "    metabolomics = {}\n",
    "\n",
    "    proteomics, transcriptomics = get_proteomics_transcriptomics_data(model, solution)\n",
    "\n",
    "    metabolomics = get_metabolomics_data(model)\n",
    "\n",
    "    return (proteomics, transcriptomics, metabolomics)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Function B\n",
    "### BEFluxes = getBEFluxes(Fluxes, Change, GSM)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getBEFluxes(fluxes_timeseries, design_df, model, sol_time_wild, tspan, delt, cell, subs, subs_ext, n_reactions=None, n_instances=None):\n",
    "    model.solver = 'cplex'\n",
    "    reaction_names = list(design_df.columns[1:])\n",
    "\n",
    "    # Calculating the number of reactions that should be modified (n_genes) and \n",
    "    # number of strains for which isoprenol concentration should be estimated \n",
    "    # mod_reax\n",
    "    if n_reactions is None:\n",
    "        n_reactions = design_df.shape[1] - 1\n",
    "    if n_instances is None:\n",
    "        n_instances = design_df.shape[0] - 1\n",
    "        \n",
    "    # series to contain the biomass optimization solution across timepoints\n",
    "    sol_time_moma = pd.Series(index=tspan)\n",
    "\n",
    "    volume = 1.0\n",
    "    # For each strain\n",
    "    for i in range(0,n_instances):\n",
    "        # At each time point\n",
    "        for t in tspan:\n",
    "            # Adding constraints to the model at each time point for each strain without globally changing the model\n",
    "            with model:\n",
    "                for k, v in subs_ext.items():\n",
    "                    model.reactions.get_by_id(k).lower_bound = max(model.reactions.get_by_id(k).lower_bound,\n",
    "                                                            -subs.loc[t,v]*volume/cell[t]/delt)\n",
    "\n",
    "                # Adding the fluxed modifications for chosen reactions\n",
    "                for reaction in reaction_names:\n",
    "                    reaction_constraint = model.problem.Constraint(model.reactions.get_by_id(reaction).flux_expression, \n",
    "                                                lb = sol_time_wild[t][reaction]*design_df.iloc[i,1],\n",
    "                                                ub = sol_time_wild[t][reaction]*design_df.iloc[i,1])\n",
    "                    model.add_cons_vars(reaction_constraint)\n",
    "\n",
    "                # Reference solution calculated for each time point in above cell for wild type\n",
    "                sol1 = sol_time_wild[t]\n",
    "\n",
    "                # Moma solution for each time point\n",
    "                sol2 = cobra.flux_analysis.moma(model, solution=sol1, linear=True) # we changed the linear=True for fadster calculation\n",
    "                \n",
    "                # saving the moma solutions across timepoints\n",
    "                sol_time_moma[t] = sol2\n",
    "                \n",
    "    return sol_time_moma\n",
    "                \n",
    "                "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Function D\n",
    "### ExtMetabolites =  Integrate(Fluxes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'10fthf_c'"
      ]
     },
     "execution_count": 183,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# def integrate_fluxes(ext_metabolites, solution_timeseries, model):\n",
    "#     '''\n",
    "#     Returns the tortal fluxes across all timepoints for all external metabolites\n",
    "#     '''\n",
    "#     extra_metabolite_fluxes = pd.Series(0, index=ext_metabolites)\n",
    "#     print(extra_metabolite_fluxes)\n",
    "#     model_metabolite_names = [met.id for met in model.metabolites]\n",
    "#     model_metabolites = [met for met in model.metabolites]\n",
    "    \n",
    "#     for metabolite in ext_metabolites:\n",
    "#         if metabolite in model_metabolite_names:\n",
    "#             reactions = list(model.metabolites.reactions)\n",
    "#         for reaction in reactions:\n",
    "            \n",
    "#         for t, solution in solution_timeseries.items():\n",
    "#             print(metabolite)\n",
    "#             print(solution)\n",
    "#             extra_metabolite_fluxes[metabolite] += solution[metabolite]\n",
    "    \n",
    "#     return extra_metabolite_fluxes\n",
    "\n",
    "# sol_time_moma\n",
    "# e_metabolites = list(user_params['ext_metabolites'].keys())\n",
    "# extra_metabolite_fluxes = integrate_fluxes(e_metabolites, sol_time_moma)\n",
    "# extra_metabolite_fluxes\n",
    "# model.metabolites[0].id"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add isopentenol pathway to model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_isopentenol_pathway(model, sce):\n",
    "    # Load S. cerevisiae model\n",
    "    #     sce = cobra.io.load_json_model(f'data/{cerevisiae_modelfile}')\n",
    "    \n",
    "    # Add mevalonate pathway reactions from S. cerevisiae model\n",
    "    for x in ['HMGCOAS','HMGCOAR','MEVK1','DPMVD']:\n",
    "        r = sce.reactions.get_by_id(x).copy()\n",
    "        r.gene_reaction_rule = ''\n",
    "        model.add_reaction(r)\n",
    "    \n",
    "    # Update gene names\n",
    "    model.reactions.get_by_id('HMGCOAS').gene_reaction_rule = 'HMGS'\n",
    "    model.reactions.get_by_id('HMGCOAR').gene_reaction_rule = 'HMGR'\n",
    "    model.reactions.get_by_id('MEVK1').gene_reaction_rule = 'MK'\n",
    "    model.reactions.get_by_id('DPMVD').gene_reaction_rule = 'PMD'\n",
    "    \n",
    "    # Add IP to model\n",
    "    m = model.metabolites.ipdp_c.copy()\n",
    "    m.id = 'ipmp_c'\n",
    "    m.name = 'Isopentenyl monophosphate'\n",
    "    m.formula = 'C5H9O4P'\n",
    "    m.charge = -2\n",
    "    model.add_metabolites([m])\n",
    "    # Update PMD reaction to convert mev-5p to IP\n",
    "    model.reactions.get_by_id('DPMVD').id = 'PMD'\n",
    "    model.reactions.get_by_id('PMD').add_metabolites({'5dpmev_c': 1.0, '5pmev_c': -1.0,\n",
    "                                                      'ipdp_c': -1.0, 'ipmp_c': 1.0})\n",
    "    # Add isoprenol (isopentenol)\n",
    "    m = model.metabolites.ipmp_c.copy()\n",
    "    m.id = 'isoprenol_c'\n",
    "    m.name = 'Isopentenol'\n",
    "    m.formula = 'C5H10O'\n",
    "    m.charge = 0\n",
    "    model.add_metabolites([m])\n",
    "    # Add phosphatase reaction by AphA\n",
    "    r = model.reactions.CHLabcpp.copy()\n",
    "    r.id = 'IPMPP'\n",
    "    r.name = 'Isopentenyl monophosphate phosphatase'\n",
    "    r.gene_reaction_rule = 'AphA'\n",
    "    model.add_reactions([r])\n",
    "    r.add_metabolites({'chol_p': 1.0, 'atp_c': 1.0, 'chol_c': -1.0, 'adp_c': -1.0, 'h_c': -1.0, 'ipmp_c': -1.0, 'isoprenol_c': 1.0})\n",
    "    \n",
    "    # Add periplasmic and extracellular isoprenol\n",
    "    m = model.metabolites.isoprenol_c.copy()\n",
    "    m.id = 'isoprenol_p'\n",
    "    m.compartment = 'p'\n",
    "    model.add_metabolites([m])\n",
    "    m = model.metabolites.isoprenol_c.copy()\n",
    "    m.id = 'isoprenol_e'\n",
    "    m.compartment = 'e'\n",
    "    model.add_metabolites([m])\n",
    "    # Add periplasmic and extracellular transport reactions\n",
    "    r = model.reactions.ETOHtrpp.copy()\n",
    "    r.id = 'IPtrpp'\n",
    "    r.name = 'Isopentenol reversible transport via diffusion (periplasm)'\n",
    "    r.gene_reaction_rule = ''\n",
    "    model.add_reactions([r])\n",
    "    r.add_metabolites({'etoh_p': 1.0, 'etoh_c': -1.0, 'isoprenol_p': -1.0, 'isoprenol_c': 1.0})\n",
    "    r = model.reactions.ETOHtex.copy()\n",
    "    r.id = 'IPtex'\n",
    "    r.name = 'Isopentenol transport via diffusion (extracellular to periplasm)'\n",
    "    r.gene_reaction_rule = ''\n",
    "    model.add_reactions([r])\n",
    "    r.add_metabolites({'etoh_e': 1.0, 'etoh_p': -1.0, 'isoprenol_e': -1.0, 'isoprenol_p': 1.0})\n",
    "    # Add a boundary reaction\n",
    "    r = model.reactions.EX_etoh_e.copy()\n",
    "    r.id = 'EX_isoprenol_e'\n",
    "    r.name = 'Isopentenol exchange'\n",
    "    r.gene_reaction_rule = ''\n",
    "    model.add_reactions([r])\n",
    "    r.add_metabolites({'etoh_e': 1.0, 'isoprenol_e': -1.0})\n",
    "    \n",
    "    # Write model to files\n",
    "    outputfilename = user_params['modelfile'].split('.')[0] + '_IPP.json'\n",
    "    cobra.io.save_json_model(model, f'data/{outputfilename}')\n",
    "    \n",
    "    return model"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Check if model has IPP pathway"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 155,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def model_has_IPP_pathway(model):\n",
    "    reaction_list = ['HMGCOAS','HMGCOAR','MEVK1','PMD','IPMPP','IPtrpp','IPtex','EX_isoprenol_e']\n",
    "    model_reactions = [r.id for r in model.reactions]\n",
    "    for reac in reaction_list:\n",
    "        if reac not in model_reactions:\n",
    "            return False\n",
    "    return True\n",
    "\n",
    "# model = cobra.io.load_json_model('iJO1366.json')\n",
    "# model_has_IPP_pathway(model)\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## User parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {},
   "outputs": [],
   "source": [
    "user_params = {\n",
    "    'host': 'ecoli', # ecoli or ropacus\n",
    "    'modelfile': 'iJO1366_MVA.json',\n",
    "    'cerevisiae_modelfile': 'iMM904.json',\n",
    "    'modelfilepath': 'sample_files',\n",
    "    'timestart': 0.0,\n",
    "    'timestop': 8.0,\n",
    "    'numpoints': 9,\n",
    "    'designfile': 'training_data_8genes.csv',\n",
    "    'designfilepath': 'data',\n",
    "    'numreactions': 8,\n",
    "    'numinstances': 5,\n",
    "    'ext_metabolites': {\n",
    "        'glc__D_e': 22.203,\n",
    "        'nh4_e': 18.695,\n",
    "        'pi_e': 69.454,\n",
    "        'so4_e': 2.0,\n",
    "        'mg2_e': 2.0,\n",
    "        'k_e': 21.883,\n",
    "        'na1_e': 103.7,\n",
    "        'cl_e': 27.25,\n",
    "        'isoprenol_e': 0.0\n",
    "    },\n",
    "    'initial_OD': 0.01\n",
    "}\n",
    "\n",
    "# comp = ['glc__D_e', 'nh4_e', 'pi_e', 'so4_e', 'mg2_e', 'k_e', 'na1_e', 'cl_e', 'EX_isoprenol_e']\n",
    "# initial_concentrations = [22.203, 18.695, 69.454, 2.0, 2.0, 21.883, 103.7, 27.25] "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Call functions\n",
    "### Function A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "metadata": {},
   "outputs": [],
   "source": [
    "file_name = user_params['modelfile']\n",
    "file_name = 'iJO1366.json'\n",
    "model = cobra.io.load_json_model(file_name)\n",
    "\n",
    "# checking if model has IPP pathway\n",
    "if not model_has_IPP_pathway(model):\n",
    "    cerevisiae_model = cobra.io.load_json_model(f'data/{user_params[\"cerevisiae_modelfile\"]}')\n",
    "    model = add_isopentenol_pathway(model, cerevisiae_model)\n",
    "\n",
    "# adding minimum flux constriants for isopentenol and formate\n",
    "iso = 'EX_isoprenol_e'\n",
    "iso_cons = model.problem.Constraint(model.reactions.EX_isoprenol_e.flux_expression,\n",
    "                        lb = 0.20)\n",
    "model.add_cons_vars(iso_cons)\n",
    "for_cons = model.problem.Constraint(model.reactions.EX_for_e.flux_expression,\n",
    "                        lb = 0.10)\n",
    "model.add_cons_vars(for_cons)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get the fluxomics time series, OD, externalk metabolites and soltion for the time series\n",
    "parameters = {\n",
    "    'timestart': user_params['timestart'],\n",
    "    'timestop': user_params['timestop'],\n",
    "    'numpoints': user_params['numpoints']\n",
    "}\n",
    "fluxomics_timeseries, od, ext_metabolites, exchange_reactions, solution_timeseries = get_flux_time_series(model, user_params['ext_metabolites'], parameters)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0.0         <Solution 0.958 at 0x7fc58e835c50>\n",
       " 1.0         <Solution 0.958 at 0x7fc588eaea10>\n",
       " 2.0         <Solution 0.958 at 0x7fc588eaefd0>\n",
       " 3.0         <Solution 0.958 at 0x7fc588eae9d0>\n",
       " 4.0         <Solution 0.958 at 0x7fc588f3e690>\n",
       " 5.0         <Solution 0.448 at 0x7fc588f3e410>\n",
       " 6.0    <Solution infeasible at 0x7fc4e502f150>\n",
       " 7.0    <Solution infeasible at 0x7fc4e4fcdad0>\n",
       " 8.0    <Solution infeasible at 0x7fc588f3e710>\n",
       " dtype: object, 0.0    0.010000\n",
       " 1.0    0.026062\n",
       " 2.0    0.067924\n",
       " 3.0    0.177026\n",
       " 4.0    0.461371\n",
       " 5.0    1.202438\n",
       " 6.0    1.881649\n",
       " 7.0    1.881649\n",
       " 8.0    1.881649\n",
       " 9.0    1.881649\n",
       " dtype: float64,     glc__D_e    nh4_e     pi_e    so4_e    mg2_e      k_e  na1_e     cl_e  \\\n",
       " 0.0   22.203   18.695   69.454        2        2   21.883  103.7    27.25   \n",
       " 1.0  22.0353  18.5215  69.4385  1.99595  1.99986  21.8799  103.7  27.2499   \n",
       " 2.0  21.5983  18.0694  69.3981  1.98539   1.9995  21.8717  103.7  27.2497   \n",
       " 3.0  20.4593   16.891  69.2929  1.95787  1.99855  21.8504  103.7  27.2491   \n",
       " 4.0  17.4909  13.8198  69.0186  1.88616  1.99608  21.7949  103.7  27.2477   \n",
       " 5.0  9.75459  5.81568  68.3037  1.69926  1.98966  21.6502  103.7  27.2438   \n",
       " 6.0        0        0  67.6485  1.52795  1.98376  21.5177  103.7  27.2403   \n",
       " 7.0        0        0  67.6485  1.52795  1.98376  21.5177  103.7  27.2403   \n",
       " 8.0        0        0  67.6485  1.52795  1.98376  21.5177  103.7  27.2403   \n",
       " 9.0        0        0  67.6485  1.52795  1.98376  21.5177  103.7  27.2403   \n",
       " \n",
       "     isoprenol_e  \n",
       " 0.0           0  \n",
       " 1.0  0.00335363  \n",
       " 2.0    0.012094  \n",
       " 3.0   0.0348733  \n",
       " 4.0   0.0942413  \n",
       " 5.0    0.248968  \n",
       " 6.0    0.552325  \n",
       " 7.0    0.552325  \n",
       " 8.0    0.552325  \n",
       " 9.0    0.552325  )"
      ]
     },
     "execution_count": 167,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "solution_timeseries, od, ext_metabolites"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Function B"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [],
   "source": [
    "# we need the solution for the wild type for maximum biomass production\n",
    "design_df = pd.read_csv(f'{user_params[\"designfilepath\"]}/{user_params[\"designfile\"]}')\n",
    "sol_time_moma = getBEFluxes(fluxomics_timeseries, design_df, model, solution_timeseries, tspan, delt, od, ext_metabolites, exchange_reactions, user_params['numreactions'], user_params['numinstances'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.0        <Solution 85.527 at 0x7fc595851b10>\n",
       "1.0        <Solution 85.527 at 0x7fc588ebd150>\n",
       "2.0        <Solution 85.527 at 0x7fc58dfeafd0>\n",
       "3.0        <Solution 85.527 at 0x7fc589c64b50>\n",
       "4.0        <Solution 85.527 at 0x7fc589bf9a90>\n",
       "5.0       <Solution 127.366 at 0x7fc588949610>\n",
       "6.0    <Solution infeasible at 0x7fc5895f0990>\n",
       "7.0    <Solution infeasible at 0x7fc58df1e210>\n",
       "8.0    <Solution infeasible at 0x7fc58dea0ad0>\n",
       "dtype: object"
      ]
     },
     "execution_count": 138,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sol_time_moma"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Function C"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tspan, delt = np.linspace(user_params['timestart'], user_params['timestop'], user_params['numpoints'], dtype='float64', retstep=True)\n",
    "proteomics_timeseries = {}\n",
    "transcriptomics_timeseries = {}\n",
    "metabolomics_timeseries = {}\n",
    "for t, solution in solution_timeseries.items():\n",
    "    proteomics_timeseries[t], transcriptomics_timeseries[t], metabolomics_timeseries[t] = get_multiomics(model, solution)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Function D"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "glc__D_e        113.541483\n",
       "nh4_e            91.812369\n",
       "pi_e            685.500041\n",
       "so4_e            17.636435\n",
       "mg2_e            19.918703\n",
       "k_e             217.000765\n",
       "na1_e          1037.000000\n",
       "cl_e            272.451222\n",
       "isoprenol_e       2.602829\n",
       "dtype: float64"
      ]
     },
     "execution_count": 143,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "integrate_fluxes(ext_metabolites)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "biodesign_3.7",
   "language": "python",
   "name": "biod_3.7"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
